index.bs

```diff
@@ -1300,253 +1300,359 @@ Promise-Using Specifications</a></strong> guideline.
-
-Both approaches to protecting against unwanted recursion have trade-offs. Some things
-to consider when choosing the guarding approach:
-
-* all state management can be rationalized within one "turn" of the algorithm (no
-    need to consider any state changes or API invocations between the time the algorithm
-    completes and when the event fires).
-* events can reveal internal state changes as early as possible when dispatched immediately
-    in the algorithm (resulting in minimal delay to user code that wants to act on
-    that state).
-* events do not need to carry snapshots (copies) of internal state because user code
-    running in the event handler can observe relevant state directly on the instance
-    object they were fired on (avoiding the need to derive a new type of {{Event}}
-    to hold such state snapshots).
-
-When deferring, event handlers will run after the algorithm ends (or starts
-to run [=in parallel=]) and any number of other tasks or microtasks may run in between
-that invalidate the object's state. Since the object's state will be unknown when
-the deferred event is dispatched, consider the following:
-
-* state relevant to the event should be packaged with the deferred event, usually
-    involving a new {{Event}}-derived type with new attributes to hold the state.
-    For example, the {{ProgressEvent}} adds {{ProgressEvent/loaded}},
-    {{ProgressEvent/total}}, etc. attributes to hold the state.
-* any coordination needed among parts of an algorithm using deferred events often
-    requires defining an explicit state machine (well-defined state transitions) to
-    ensure that when a deferred event fires, the behavior of inspecting or changing
-    state is well-defined. For example, in [[payment-request]], the {{PaymentRequest}}'s
-    [=[[state]]=] internal slot explicitly tracks the object's state
-    through its well-defined transitions.
-    * in addition to defining state transitions, each coordinated algorithm usually
-        applies the guarding technique (anyway) to ensure the algorithm can only proceed
-        under the appropriate set of states. For example, in [[payment-request]] note
-        the guards used often around the [=[[state]]=] internal
-        slot such as in the {{MerchantValidationEvent/complete()}} algorithm.
-
-Finally, dispatching a deferred event that does not seem to require
-packaging extra state or defining a state machine for the algorithm as mentioned above,
-could mean that all of the state transitions have been completed and that the event
-is meant to signal completion of the algorithm. In this case, it's likely that instead
-of using an event to signal completion, the API should be designed to return and complete
-a {{Promise}} instead. See [[#one-time-events]].
+If using events causes problems,
+such as unavoidable [[#guard-against-recursion|recursion]],
+consider using an Observer pattern instead.
```

Comment:
link?

Reply:
I think the link would go to this section...